File: code/packages/go/bezier2d/bezier2d_test.go
package bezier2d

import (
	"math"
	"testing"

	"github.com/adhithyan15/coding-adventures/code/packages/go/point2d"
)

const eps = 1e-9

func approxEq(a, b float64) bool { return math.Abs(a-b) < eps }
func ptEq(a, b point2d.Point) bool {
	return approxEq(a.X, b.X) && approxEq(a.Y, b.Y)
}

var q = QuadraticBezier{
	point2d.NewPoint(0, 0), point2d.NewPoint(1, 2), point2d.NewPoint(2, 0),
}

func TestQuadEndpoints(t *testing.T) {
	if !ptEq(EvalQuad(q, 0), q.P0) {
		t.Error("EvalQuad at t=0 should be P0")
	}
	if !ptEq(EvalQuad(q, 1), q.P2) {
		t.Error("EvalQuad at t=1 should be P2")
	}
}

func TestQuadMidpoint(t *testing.T) {
	m := EvalQuad(q, 0.5)
	if !approxEq(m.X, 1) || !approxEq(m.Y, 1) {
		t.Errorf("QuadMidpoint: got %v", m)
	}
}

func TestQuadSplitEndpoints(t *testing.T) {
	left, right := SplitQuad(q, 0.5)
	m := EvalQuad(q, 0.5)
	if !ptEq(left.P2, m) || !ptEq(right.P0, m) {
		t.Error("Split midpoints don't match")
	}
}

func TestQuadPolylineStraight(t *testing.T) {
	straight := QuadraticBezier{
		point2d.NewPoint(0, 0), point2d.NewPoint(1, 0), point2d.NewPoint(2, 0),
	}
	pts := PolylineQuad(straight, 0.1)
	if len(pts) != 2 {
		t.Errorf("Straight line should give 2 points, got %d", len(pts))
	}
}

func TestQuadBboxContainsEndpoints(t *testing.T) {
	bb := BboxQuad(q)
	if bb.X > 0 || bb.X+bb.Width < 2 {
		t.Errorf("BboxQuad doesn't contain endpoints: %v", bb)
	}
}

func TestQuadElevateEquivalent(t *testing.T) {
	c := ElevateQuad(q)
	for _, t_ := range []float64{0, 0.25, 0.5, 0.75, 1} {
		qp := EvalQuad(q, t_)
		cp := EvalCubic(c, t_)
		if !approxEq(qp.X, cp.X) || !approxEq(qp.Y, cp.Y) {
			t.Errorf("ElevateQuad mismatch at t=%v: %v vs %v", t_, qp, cp)
		}
	}
}

var c = CubicBezier{
	point2d.NewPoint(0, 0), point2d.NewPoint(1, 2),
	point2d.NewPoint(3, 2), point2d.NewPoint(4, 0),
}

func TestCubicEndpoints(t *testing.T) {
	if !ptEq(EvalCubic(c, 0), c.P0) {
		t.Error("EvalCubic at t=0 should be P0")
	}
	if !ptEq(EvalCubic(c, 1), c.P3) {
		t.Error("EvalCubic at t=1 should be P3")
	}
}

func TestCubicSymmetricMidpoint(t *testing.T) {
	m := EvalCubic(c, 0.5)
	if !approxEq(m.X, 2) {
		t.Errorf("Symmetric midpoint X should be 2, got %v", m.X)
	}
}

func TestCubicSplitEndpoints(t *testing.T) {
	left, right := SplitCubic(c, 0.5)
	m := EvalCubic(c, 0.5)
	if !ptEq(left.P3, m) || !ptEq(right.P0, m) {
		t.Error("Split midpoints don't match")
	}
}

func TestCubicPolylineStraight(t *testing.T) {
	straight := CubicBezier{
		point2d.NewPoint(0, 0), point2d.NewPoint(1, 0),
		point2d.NewPoint(2, 0), point2d.NewPoint(3, 0),
	}
	pts := PolylineCubic(straight, 0.1)
	if len(pts) != 2 {
		t.Errorf("Straight line should give 2 points, got %d", len(pts))
	}
}

func TestCubicBboxContainsSamples(t *testing.T) {
	bb := BboxCubic(c)
	for i := 0; i <= 20; i++ {
		p := EvalCubic(c, float64(i)/20)
		if p.X < bb.X-1e-6 || p.X > bb.X+bb.Width+1e-6 {
			t.Errorf("X out of bbox at i=%d: %v", i, p)
		}
	}
}

File: code/packages/elixir/bezier2d/test/bezier2d_test.exs
defmodule Bezier2DTest do
  use ExUnit.Case

  @delta 1.0e-9
  defp approx(a, b), do: abs(a - b) < @delta
  defp pt_approx({x1, y1}, {x2, y2}), do: approx(x1, x2) and approx(y1, y2)

  @q Bezier2D.new_quad({0, 0}, {1, 2}, {2, 0})

  test "quad eval at 0" do
    assert pt_approx(Bezier2D.eval_quad(@q, 0), {0, 0})
  end

  test "quad eval at 1" do
    assert pt_approx(Bezier2D.eval_quad(@q, 1), {2, 0})
  end

  test "quad eval midpoint" do
    assert pt_approx(Bezier2D.eval_quad(@q, 0.5), {1, 1})
  end

  test "quad split midpoints match" do
    {left, right} = Bezier2D.split_quad(@q, 0.5)
    m = Bezier2D.eval_quad(@q, 0.5)
    assert pt_approx(elem(left, 3), m)
    assert pt_approx(elem(right, 1), m)
  end

  test "quad polyline straight" do
    straight = Bezier2D.new_quad({0, 0}, {1, 0}, {2, 0})
    pts = Bezier2D.polyline_quad(straight, 0.1)
    assert length(pts) == 2
  end

  test "quad bbox contains endpoints" do
    {rx, _ry, rw, _rh} = Bezier2D.bbox_quad(@q)
    assert rx <= 0 and rx + rw >= 2
  end

  test "quad elevate equivalent" do
    c = Bezier2D.elevate_quad(@q)
    Enum.each([0, 0.25, 0.5, 0.75, 1], fn t ->
      qp = Bezier2D.eval_quad(@q, t)
      cp = Bezier2D.eval_cubic(c, t)
      assert pt_approx(qp, cp)
    end)
  end

  @c Bezier2D.new_cubic({0, 0}, {1, 2}, {3, 2}, {4, 0})

  test "cubic eval at 0" do
    assert pt_approx(Bezier2D.eval_cubic(@c, 0), {0, 0})
  end

  test "cubic eval at 1" do
    assert pt_approx(Bezier2D.eval_cubic(@c, 1), {4, 0})
  end

  test "cubic symmetric midpoint x" do
    {x, _} = Bezier2D.eval_cubic(@c, 0.5)
    assert approx(x, 2)
  end

  test "cubic split midpoints match" do
    {left, right} = Bezier2D.split_cubic(@c, 0.5)
    m = Bezier2D.eval_cubic(@c, 0.5)
    assert pt_approx(elem(left, 4), m)
    assert pt_approx(elem(right, 1), m)
  end

  test "cubic polyline straight" do
    straight = Bezier2D.new_cubic({0, 0}, {1, 0}, {2, 0}, {3, 0})
    pts = Bezier2D.polyline_cubic(straight, 0.1)
    assert length(pts) == 2
  end

  test "cubic bbox contains samples" do
    {bx, by, bw, bh} = Bezier2D.bbox_cubic(@c)
    Enum.each(0..20, fn i ->
      {px, _py} = Bezier2D.eval_cubic(@c, i / 20.0)
      assert px >= bx - 1.0e-6 and px <= bx + bw + 1.0e-6
    end)
  end
end
